client/network/src/service.rs

```diff
@@ -1027,6 +1042,34 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
 			.to_worker
 			.unbounded_send(ServiceToWorkerMsg::OwnBlockImported(hash, number));
 	}
+
+	/// Send a bitswap block to a peer.
+	pub fn bitswap_send_block(&self, peer_id: PeerId, cid: tiny_cid::Cid, data: Box<[u8]>) {
+		let _ = self
+			.to_worker
+			.unbounded_send(ServiceToWorkerMsg::BitswapSendBlock(peer_id, cid, data));
+	}
+
+	/// Send a bitswap block to all peers that have the block in their wantlist.
+	pub fn bitswap_send_block_all(&self, cid: tiny_cid::Cid, data: Box<[u8]>) {
+		let _ = self
+			.to_worker
+			.unbounded_send(ServiceToWorkerMsg::BitswapSendBlockAll(cid, data));
+	}
+
+	/// Send a bitswap WANT request to all peers for a block.
+	pub fn bitswap_want_block(&self, cid: tiny_cid::Cid, priority: i32) {
+		let _ = self
+			.to_worker
+			.unbounded_send(ServiceToWorkerMsg::BitswapWantBlock(cid, priority));
+	}
+
+	/// Cancel a bitswap WANT request.
+	pub fn bitswap_cancel_block(&self, cid: tiny_cid::Cid) {
+		let _ = self
+			.to_worker
+			.unbounded_send(ServiceToWorkerMsg::BitswapCancelBlock(cid));
+	}
```

Comment:
This needs to be documented.
I don't know how bitswap works, and that's great because it should in my opinion be possible for someone who doesn't know how bitswap works to understand this API.

What's the general workflow of the thing? What happens if `send_block` is called on a peer that doesn't want this block?

Reply:
bitswap_send_block is usually in response to a want event. the block is added to a queue and if a cancel request is received before the message is sent it will be removed from the queue. reading the code, if the cancel request comes in before the want request was processed, the cancel request will be ignored. I guess it could use some work, but it's not "terrible"